### PR TITLE
Fixes VSTS 896684: Override keyword does not show protocol completions

### PR DESCRIPTION
**Pull request by avodovnik**
Setting IsExclusive to true and ensuring our `ProtocolCompletionMemberProvider` gets called even after `override` keyword is already written. 

This change will ensure our completion data will get surfaced *alongisde* completion suggestions from the other exclusive provider, which is the actual override provider. There was a recent Roslyn change, starting from 16.2.P4 where exclusivity simply means grouping - showing all exclusive ones, not just whoever gets there first. Additionally, we are now ensuring our provider also gets called when there is an insertion/deletion going on (i.e. after the user writes override).

Note: This should only start actually working when we get the Roslyn change that includes the merge from [this PR](https://github.com/dotnet/roslyn/pull/36088), specifically commit: `3a364d2a2600c4a61b2ac7e68c1ac380dc07164a`. I've added the `do-not-merge` label onto it, because it's impossible to properly _test_ before that change. While I've checked, and it _should_ not have any adverse effects, it's probably worth being prudent. 

VSTS: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/896684

### CODE REVIEW

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs [hunk 1]
```diff
@@ -468,5 +483,23 @@ static bool TryDetermineModifiers (ref SyntaxToken startToken, SourceText text,
 					.WithIsSealed (isSealed);*/
 			return overrideToken.IsKind (SyntaxKind.OverrideKeyword) && IsOnStartLine (text, startLine, overrideToken.Parent.SpanStart);
 		}
+
+		public override bool ShouldTriggerCompletion (SourceText text, int position, CompletionTrigger trigger, Microsoft.CodeAnalysis.Options.OptionSet options)
```
Comment (sandyarmstrong) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs]: Have you verified that behavior is still correct in old editor with this additional change?

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs [hunk 2]
```diff
@@ -97,6 +97,21 @@ public override async Task ProvideCompletionsAsync (Microsoft.CodeAnalysis.Compl
 					return;
 
 				var startToken = token.GetPreviousTokenIfTouchingWord (position);
+
+				// We're marking the context as exclusive, to resolve the regression
+				// that appeared since introducing Roslyn's provider mechanism. The
```
Comment (sandyarmstrong) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs]: Maybe specifically mention here that this is a regression only in the new editor, which interacts with Roslyn completion differently than the old editor.

**Comment by sandyarmstrong** [comment]
Relevant Roslyn update here: https://github.com/mono/monodevelop/pull/8019

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs [hunk 3]
```diff
@@ -97,6 +97,21 @@ public override async Task ProvideCompletionsAsync (Microsoft.CodeAnalysis.Compl
...
+				// actual override completion provider took precdence and therefore
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs]: `precdence` -> `precedence`

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs [hunk 4]
```diff
@@ -468,5 +483,23 @@ static bool TryDetermineModifiers (ref SyntaxToken startToken, SourceText text,
 					.WithIsSealed (isSealed);*/
 			return overrideToken.IsKind (SyntaxKind.OverrideKeyword) && IsOnStartLine (text, startLine, overrideToken.Parent.SpanStart);
 		}
+
+		public override bool ShouldTriggerCompletion (SourceText text, int position, CompletionTrigger trigger, Microsoft.CodeAnalysis.Options.OptionSet options)
+		{
+			// This method is overriden because Roslyn's Completion Service calls us, in cases
```
Comment (KirillOsenkov) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs]: `overridden`

**Comment by sandyarmstrong** [comment]
Works as expected with Roslyn bump in https://github.com/mono/monodevelop/pull/8019.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs (hunk 1 above)
Comment (sandyarmstrong) [reply main/src/addins/CSharpBinding/MonoDevelop.CSharp.Completion/CompletionProvider/ProtocolMemberCompletionProvider.cs]: I have tested in the old editor and we seem good here.

**Comment by sandyarmstrong** [comment]
Fixed all comments, rebased, removed `do-not-merge` label. This is good to go now!

**Comment by abock** [comment]
@monojenkins backport release-8.2

**Pull request merged**